data_scrapers/alameda_county.py

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env python3
+import requests
+import urllib.request
+from bs4 import BeautifulSoup
+import json
+import pandas as pd
+from typing import List, Dict
+from datetime import datetime, timedelta, timezone
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+import re
+import os
+
+# Note that we are using numbers for all of Alameda County, including Berkeley
+# Running this scraper requires a Firefox webdriver. The macos Firefox driver, geckodriver, is stored in ./env/bin
+# This is the open data landing page: https://data.acgov.org/search?source=alameda%20county%20public%20health%20department&tags=covid-19
+    # Getting strted with the ArcGIS REST API: https://developers.arcgis.com/rest/services-reference/get-started-with-the-services-directory.htm
+    # Services reference for Layer: https://developers.arcgis.com/rest/services-reference/layer-feature-service-.htm
+
+# URLs and API endpoints:
+landing_page = "https://data.acgov.org/search?source=alameda%20county%20public%20health%20department&tags=covid-19"
+cases_deaths = 'https://opendata.arcgis.com/datasets/7ea4fd9b8a1040a7b3815f2e0b5f92ba_0/FeatureServer/0/query'
+demographics = 'https://opendata.arcgis.com/datasets/f218564293f9400a8296558e9325f265_0/FeatureServer/0/query'
+cases_meta = 'https://services3.arcgis.com/1iDJcsklY3l3KIjE/arcgis/rest/services/AC_dates/FeatureServer/0?f=json'
+demographics_meta = 'https://services3.arcgis.com/1iDJcsklY3l3KIjE/arcgis/rest/services/AC_geography/FeatureServer/0?f=json'
+dashboards = ['https://ac-hcsa.maps.arcgis.com/apps/opsdashboard/index.html#/1e0ac4385cbe4cc1bffe2cf7f8e7f0d9',
+              'https://ac-hcsa.maps.arcgis.com/apps/opsdashboard/index.html#/332a092bbc3641bd9ec8373e7c7b5b3d']
+
+# Load data model template into the 'out' dictionary. 'out' will be global to all methods.
+with open('./data_scrapers/_data_model.json') as template:
+    out = json.load(template)
```

Comment:
Agree that using a global is not great. I've moved it into get_county(), and it will return a fresh object by opening the file when called. 
Re: using this as a template vs. a reference, I thought it would be a good way for front-end and other users to see data that is not being collected by the county, since those values would have the default template of "-1". I also thought it would be a good centralized location to change and add labels and data-points if necessary. But maybe it's better to use it as a reference only, or at least grab only the elements of the template for which our sources actually provide data.

Reply:
> I thought it would be a good way for front-end and other users to see data that is not being collected by the county, since those values would have the default template of "-1". I also thought it would be a good centralized location to change and add labels and data-points if necessary.

Both of these reasons totally make sense! 👍

> maybe it's better to… grab only the elements of the template for which our sources actually provide data.

This might be fine, too. **I think what you’ve got now is totally OK.** If you really wanted to dig in more, something like this might be more flexible:

```py
from functools import lru_cache
from pathlib import Path

# lru_cache will memoize the function. You could also do this manually ¯\_(ツ)_/¯
@lru_cache()
def json_file(filepath):
    with open(filepath) as file:
        return json.load(file)

def county_template():
    # Get path relative to this file, so we don't need to worry
    # about what working directory the script was run from.
    template_path = Path(__file__).parent / '_data_model.json'
    return copy.deepcopy(json_file(template_path))

def get_county() -> Dict:
    # Get a copy of the template to fill in
    out = county_template()

# This version of the function doesn't need the template as input.
# Since the file reading and parsing is cached, it's pretty low
# overhead to ask for a fresh copy (we're just making a copy of data
# already in memory).
def get_demographics() -> (Dict, List):
    out = county_template()['field_you_care_about']
```

But really, what you’ve got now is a-ok. 👌 Let’s not worry *too* much here and get this out the door.